cmd/repo-query/repo-query.go

```diff
@@ -0,0 +1,353 @@
+	}
+}
+
+type indexPostRequest struct {
+	Owner  string `json:"owner"`
+	Name   string `json:"name"`
+	Branch string `json:"branch"`
+}
+
+func indexRepo(owner string, repo string, branch string) error {
+	indexPostReq := &indexPostRequest{
+		Owner:  owner,
+		Name:   repo,
+		Branch: branch,
+	}
+
+	indexPostJSON, err := json.Marshal(indexPostReq)
+	if err != nil {
+		return err
+	}
+
+	responseBody := bytes.NewBuffer(indexPostJSON)
+	resp, err := http.Post(fmt.Sprintf("%s/embed", repoQueryURL), "application/json", responseBody)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("An error occurred: %v\n", string(body))
+		return errors.New("error while indexing repository")
+	}
+
+	// listen for SSEs and send data,event pairs to processIndexChunk
+	reader := bufio.NewReader(resp.Body)
+	for {
+		line, err := reader.ReadString('\n')
+		// if err is eof then break
+		if err == io.EOF {
+			return nil
+		} else if err != nil {
+			return err
+		}
+
+		if line == "\n" {
+			continue
+		}
+
+		if strings.HasPrefix(line, "event: ") {
+			chunk := line
+			for {
+				line, err := reader.ReadString('\n')
+				if err != nil {
+					return err
+				}
+
+				if line == "\n" {
+					break
+				}
+
+				chunk += line
+			}
+
+			err := processIndexChunk(chunk)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func processIndexChunk(chunk string) error {
+	chunkLines := strings.Split(chunk, "\n")
+	eventLine := chunkLines[0]
+	event := strings.Split(eventLine, ": ")[1]
+
+	switch event {
+	case "FETCH_REPO":
+		fmt.Println("Fetching Repository from GitHub...")
+	case "EMBED_REPO":
+		fmt.Println("Embedding Repository...")
+	case "SAVE_EMBEDDINGS":
+		fmt.Println("Saving the embeddings to our database...")
+	case "ERROR":
+		fmt.Println("There was an error while indexing this repository.")
+		return errors.New("error while indexing repository")
+	case "DONE":
+		fmt.Println("Indexing Complete. You can now ask questions about this repository! 🎉")
+	default:
+		break
+	}
+
+	return nil
+}
+
+type queryPostRequest struct {
+	Query      string `json:"query"`
+	Repository struct {
+		Owner  string `json:"owner"`
+		Name   string `json:"name"`
+		Branch string `json:"branch"`
+	} `json:"repository"`
+}
+
+func askQuestion(question string, owner string, repo string, branch string) error {
+	queryPostReq := &queryPostRequest{
+		Query: question,
+		Repository: struct {
+			Owner  string `json:"owner"`
+			Name   string `json:"name"`
+			Branch string `json:"branch"`
+		}{
+			Owner:  owner,
+			Name:   repo,
+			Branch: branch,
+		},
+	}
+
+	queryPostJSON, err := json.Marshal(queryPostReq)
+	if err != nil {
+		return err
+	}
+
+	responseBody := bytes.NewBuffer(queryPostJSON)
+	resp, err := http.Post(fmt.Sprintf("%s/query", repoQueryURL), "application/json", responseBody)
```

Comment:
Same here: could simplify by using the buffer directly:

```suggestion
	resp, err := http.Post(fmt.Sprintf("%s/query", repoQueryURL), "application/json", bytes.NewBuffer(queryPostJSON))
```